### PR TITLE
SNAPSHOT: Introduce ClusterState Tests

### PR DESCRIPTION
**Pull request by original-brownbear**
* Bringing in cluster state test infrastructure
* Relates #32265

### CODE REVIEW

**Comment by original-brownbear** [comment]
Jenkins test this

**Comment by original-brownbear** [comment]
Jenkins test this

**Comment by original-brownbear** [comment]
Jenkins test this

File: server/src/test/java/org/elasticsearch/indices/cluster/ClusterStateChanges.java [hunk 1]
```diff
@@ -291,14 +364,21 @@ public ClusterState applyStartedShards(ClusterState clusterState, List<ShardRout
     }
 
     private ClusterState executeClusterStateUpdateTask(ClusterState state, Runnable runnable) {
-        ClusterState[] result = new ClusterState[1];
+        when(clusterService.state()).thenReturn(state);
+        CompletableFuture<ClusterState> clusterStateResult = new CompletableFuture<>();
         doAnswer(invocationOnMock -> {
             ClusterStateUpdateTask task = (ClusterStateUpdateTask)invocationOnMock.getArguments()[1];
-            result[0] = task.execute(state);
+            assertTrue(clusterStateResult.complete(task.execute(state)));
```
Comment (original-brownbear) [inline server/src/test/java/org/elasticsearch/indices/cluster/ClusterStateChanges.java]: I changed this to allow us to get the result of an asynchronously completing action triggered by the `runnable` and added the assertion to make sure we're only ever handling a single state update here.

**Comment by original-brownbear** [comment]
@ywelsch how about this for the next step in this effort? Also, I wonder if we actually have to limit this to the feature branch. There's only a visibility change as far as production code changes go, maybe fine to target `master` here?

File: server/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java [hunk 2]
```diff
@@ -99,6 +107,44 @@ public void tearDown() throws Exception {
...
+        RuntimeException ex =
+            expectThrows(RuntimeException.class, () -> cluster.deleteIndices(finalState, new DeleteIndexRequest("test-index")));
+        assertEquals(
+            "Cannot delete indices that are being snapshotted: [[test-index]]. " +
+                "Try again after snapshot finishes or cancel the currently running snapshot.",
+            ex.getCause().getMessage()
+        );
+        state = cluster.deleteSnapshot(state, new DeleteSnapshotRequest("test-repo", "test-snap"));
+        assertNotNull(state.custom(SnapshotsInProgress.TYPE));
+        // TODO: This should be 1, but we're currently leaving one aborted entry behind
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/indices/cluster/IndicesClusterStateServiceRandomUpdatesTests.java]: what would it take to fix this? What other tests could we possibly add here? Move this tests (and other new ones) to their own test class?

**Comment by original-brownbear** [comment]
@ywelsch I extracted the tests to their own class. Let me know if this is an ok step :)

**Comment by original-brownbear** [comment]
Closing in favor of https://github.com/elastic/elasticsearch/pull/36644

**Pull request closed without merging**